packages/@headlessui-react/src/components/combobox/combobox.tsx

```diff
@@ -1689,8 +1733,24 @@ function OptionFn<
     /* We also want to trigger this when the position of the active item changes so that we can re-trigger the scrollIntoView */ data.activeOptionIndex,
   ])
 
-  let handleClick = useEvent((event: { preventDefault: Function }) => {
-    if (disabled || data.virtual?.disabled(value)) return event.preventDefault()
+  let handleMouseDown = useEvent((event: ReactMouseEvent<HTMLButtonElement>) => {
+    // We use the `mousedown` event because this will fire before the `focus`
+    // event. When we use `event.preventDefault()` here, the
+    // `document.activeElement` will stay on the current element.
+    //
+    // Typically that means that the `ComboboxInput` will stay focused, and thus
+    // the selection / cursor position will be preserved.
+    event.preventDefault()
+
+    // When we used the `click` event we didn't have to worry about this because
+    // that only fires if you use the `left` mouse button. However, now that we
+    // use the `mousedown` event, we do have to worry about which button is
+    // pressed.
+    if (event.button !== MouseButton.Left) {
+      return
+    }
+
+    if (disabled || data.virtual?.disabled(value)) return
```

Comment:
See previous comment suggestions.

Also, this feels like some duplication — I'm sure you've done that for good reason, but should we consider moving this into an abstraction? 🤔

Reply:
Yep 2 separate locations, didn't want to create an abstraction just yet. Maybe this is something that could live in a hook but didn't want to think about a name yet.

Reply:
Makes sense!